streamalert/rules_engine/rules_engine.py

```diff
@@ -213,15 +213,82 @@ def _rule_analysis(self, payload, rule):
 
         return alert
 
-    def _configure_outputs(self, rule):
+    def _configure_outputs(self, record, rule):
+        """Configure the outputs for the rule
+
+        Args:
+            record (dict): Record to pass through to dynamic_outputs
+            rule (rule.Rule): Attributes for the rule which triggered the alert
+        Returns:
+            set: unique set of outputs, only required outputs if the rule is staged
+        """
         # Check if the rule is staged and, if so, only use the required alert outputs
         if rule.is_staged(self._rule_table):
             all_outputs = self._required_outputs_set
         else:  # Otherwise, combine the required alert outputs with the ones for this rule
             all_outputs = self._required_outputs_set.union(rule.outputs_set)
+            if rule.dynamic_outputs:
+                dynamic_outputs = self._configure_dynamic_outputs(record, rule)
+                all_outputs = all_outputs.union(dynamic_outputs)
```

Comment:
Nitpick; I recommend eliminating duplicates from `all_outputs`. Maybe return a Python set from `dynamic_outputs`, or if we use a list, we could dedupe with something like 

```
all_outputs = list(set(all_outputs) | set(dynamic_outputs))
```

or something

Reply:
I did struggle with how to add the outputs bit here, wasn't sure what to call it. Since I need to take three sets of outputs and create one unique set (`required_outputs`, `rule.output_set` and `rule.dynamic_outputs`) prehaps there's a cleaner way of combining like you mentioned?

Reply:
If each of the inputs is a `set` you can use a three way `union`

```
set1.union(set2, set3)
```

The thing is I'm pretty sure not all of the three of them are `set`s.

Reply:
take another look at how i configured `_configure_outputs` i personally think this new implementation is better as its not trying to do what i had before. Its going through each `output_source` and then adding the outputs to a set. This made testing this much easier!